llvm-external-projects/iree-dialects/lib/Dialect/IREEPyDM/Transforms/Rtl/LowerToRtlPass.cpp

```diff
@@ -0,0 +1,170 @@
+    rewriter.create<pydm_d::RaiseOnFailureOp>(loc, callOp.exc_result());
+    return callOp.result();
+  }
+
+  void replaceOpWithCall(Operation *op, ValueRange inputs,
+                         PatternRewriter &rewriter) const {
+    Value callResult = emitImportCall(op->getLoc(), inputs, rewriter);
+    assert(op->getNumResults() != 0 && "expected op with results");
+    if (op->getNumResults() == 1) {
+      // No unpack.
+      rewriter.replaceOp(op, {callResult});
+    } else {
+      // Unpack 1 -> N.
+      SmallVector<Type> unpackTypes = {
+          rewriter.getType<pydm_d::ExceptionResultType>()};
+      unpackTypes.append(op->getResultTypes().begin(),
+                         op->getResultTypes().end());
+      auto unpackOp = rewriter.create<pydm_d::DynamicUnpackOp>(
+          op->getLoc(), unpackTypes, callResult);
+      rewriter.create<pydm_d::RaiseOnFailureOp>(op->getLoc(),
+                                                unpackOp.exc_result());
+      rewriter.replaceOp(op, unpackOp.slots());
+    }
+  }
+
+ private:
+  SymbolTable &symbolTable;
+};
+
+struct ObjectAsBoolPattern
+    : public EmitImportCallBase<ObjectAsBoolFunc, pydm_d::AsBoolOp> {
+  using EmitImportCallBase::EmitImportCallBase;
+
+  LogicalResult matchAndRewrite(pydm_d::AsBoolOp srcOp,
+                                PatternRewriter &rewriter) const override {
+    auto valueType = srcOp.value().getType().dyn_cast<pydm_d::ObjectType>();
+    if (!valueType)
+      return rewriter.notifyMatchFailure(srcOp, "not an !object<>");
+    replaceOpWithCall(srcOp, {srcOp.value()}, rewriter);
+    return success();
+  }
+};
+
+struct DynamicBinaryPromotePattern
+    : public EmitImportCallBase<DynamicBinaryPromoteFunc,
+                                pydm_d::DynamicBinaryPromoteOp> {
+  using EmitImportCallBase::EmitImportCallBase;
+
+  LogicalResult matchAndRewrite(pydm_d::DynamicBinaryPromoteOp srcOp,
+                                PatternRewriter &rewriter) const override {
+    replaceOpWithCall(srcOp, {srcOp.left(), srcOp.right()}, rewriter);
+    return success();
+  }
+};
+
+struct LowerIREEPyDMToRtlPass
+    : public LowerIREEPyDMToRtlBase<LowerIREEPyDMToRtlPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry
+        .insert<mlir::iree::IREEDialect, BuiltinDialect, StandardOpsDialect>();
+  }
+
+  void runOnOperation() override {
+    auto *context = &getContext();
+    auto moduleOp = getOperation();
+    SymbolTable symbolTable(moduleOp);
+    RewritePatternSet patterns(context);
+    patterns.insert<DynamicBinaryPromotePattern, ObjectAsBoolPattern>(
+        symbolTable);
+
+    GreedyRewriteConfig config;
+    if (failed(applyPatternsAndFoldGreedily(moduleOp, std::move(patterns),
+                                            config))) {
+      emitError(getOperation().getLoc())
+          << "did not converge while lowering to rtl";
+      signalPassFailure();
```

Comment:
a trick here is that return is fine with void returns; so you can do `return signalPassFailure();` - this prevents accidental fallthrough if someone added code below this if()